File: JAVA/962. Maximum Width Ramp.java
class Solution {
    public int maxWidthRamp(int[] nums) {
        int len = nums.length;
        int[] leftmin = new int[len];
        int min = nums[0];
        for(int i=0;i<len;++i)
        {
            if(nums[i]<min)
                min = nums[i];
            leftmin[i] = min;
        }
        int[] rightmax = new int[len];
        int max = nums[len-1];
        for(int i=len-1;i>=0;--i)
        {
            if(nums[i]>max)
                max = nums[i];
            rightmax[i] = max;
        }
        int maxLength = -1;
        int i=0,j=0;
        while(i<=j && i<len && j<len)
        {
            if(leftmin[i]<=rightmax[j])
            {
                maxLength = Math.max(maxLength,j-i);
                j++;
            }
            else
                i++;
        }
        return maxLength;
    }
}

File: C++/962. Maximum Width Ramp.cpp
class Solution {
public:
    int maxWidthRamp(vector<int>& nums) {
        int n = nums.size();
        stack<int> s;
        for (int i = 0; i < n; i++) {
            if (s.empty() || nums[i] < nums[s.top()]) {
                s.push(i);
            }
        }        
        int ans = 0;
        for (int j = n - 1; j >= 0; j--) {
            while (!s.empty() && nums[s.top()] <= nums[j]) {
                ans = max(ans, j - s.top());
                s.pop();
            }
        }
        return ans;
    }
};
